source/distributed/cell_weights.cc

```diff
@@ -63,71 +44,186 @@ namespace parallel
-      const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)>
-      custom_function)
+  template <typename FactorType, typename ExponentType>
+  typename CellWeights<dim, spacedim>::WeightingFunction
+  CellWeights<dim, spacedim>::ndofs_weighting(
+    const std::vector<std::pair<FactorType, ExponentType>> &coefficients)
   {
-    weighting_function = custom_function;
+    if (std::is_same<ExponentType, unsigned int>::value)
+      {
+        return [&coefficients](
+                 const typename hp::DoFHandler<dim, spacedim>::cell_iterator &,
+                 const FiniteElement<dim, spacedim> &future_fe)
+                 -> unsigned int {
+          const typename std::common_type<FactorType, unsigned int>::type
+            result = 0;
+          for (const auto &pair : coefficients)
+            result +=
+              pair.first * Utilities::pow(future_fe.dofs_per_cell, pair.second);
+          result = std::trunc(result);
+
+          Assert(
+            result <= std::numeric_limits<unsigned int>::max(),
+            ExcMessage(
+              "Cannot cast determined weight for this cell to unsigned int!"));
+
+          return static_cast<unsigned int>(result);
+        };
+      }
+    else
+      {
+        return [&coefficients](
+                 const typename hp::DoFHandler<dim, spacedim>::cell_iterator &,
+                 const FiniteElement<dim, spacedim> &future_fe)
+                 -> unsigned int {
+          const typename std::common_type<FactorType,
+                                          ExponentType,
+                                          unsigned int>::type result = 0;
+          for (const auto &pair : coefficients)
+            result +=
+              pair.first * std::pow(future_fe.dofs_per_cell, pair.second);
+          result = std::trunc(result);
+
+          Assert(
+            result <= std::numeric_limits<unsigned int>::max(),
+            ExcMessage(
+              "Cannot cast determined weight for this cell to unsigned int!"));
+
+          return static_cast<unsigned int>(result);
+        };
+      }
+  }
+
+
+
+  // ---------- handling callback functions ----------
+
+  template <int dim, int spacedim>
+  std::function<unsigned int(
+    const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+    const typename Triangulation<dim, spacedim>::CellStatus     status)>
+  CellWeights<dim, spacedim>::make_weighting_callback(
+    const hp::DoFHandler<dim, spacedim> &dof_handler,
+    const typename CellWeights<dim, spacedim>::WeightingFunction
+      &weighting_function)
+  {
+    const parallel::TriangulationBase<dim, spacedim> *tria =
+      dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+        &(dof_handler.get_triangulation()));
+
+    Assert(
+      tria != nullptr,
+      ExcMessage(
+        "parallel::CellWeights requires a parallel::TriangulationBase object."));
+
+    return std::bind(&CellWeights<dim, spacedim>::weighting_callback,
```

Comment:
@masterleinad has been working hard to banish `std::bind` from the code base by using lambda functions instead -- he's going to stop liking you if you introduce new ones ;-)